codegen/vector/main.py

```diff
@@ -0,0 +1,692 @@
+"""
+
+import os
+import sys
+import math
+from constants import VLEN
+from templates import (
+    TEMPLATE_HEADER,
+    VLRE_TEMPLATE,
+    MAKEFRAG_TEMPLATE,
+    VLE_CODE_TEMPLATE,
+    VLSE_TEMPLATE,
+    VSE_CODE_TEMPLATE,
+    MASK_CODE,
+    ARITH_VF_CODE_TEMPLATE,
+    ARITH_VI_CODE_TEMPLATE,
+    ARITH_VV_CODE_TEMPLATE,
+    ARITH_VX_CODE_TEMPLATE,
+    ARITH_TEMPLATE,
+)
+from utils import (
+    byte_masked,
+    generate_test_data,
+    inc,
+    get_element,
+    merge_quads,
+    save_to_file,
+    lshift,
+    rshift,
+    floathex,
+    cast_insts,
+)
+
+
+class CodegenBase:
+    """Codegen base class."""
+
+    class Option:
+        """Codegen base option."""
+
+        def __init__(self, filename):
+            self.filename = filename
+
+        def nbytes(self):
+            """Number of bytes for data needs to process."""
+            return 0
+
+    def __str__(self):
+        return "Not implemented."
+
+
+class Vlre(CodegenBase):
+    """Generate vl<NF>re<EEW>.v tests."""
+
+    class Option(CodegenBase.Option):
+        """Vlre option."""
+
+        def __init__(self, filename, nf, eew):
+            super().__init__(filename)
+            self.nf = nf
+            self.eew = eew
+
+        def nbytes(self):
+            return (VLEN // 8) * self.nf
+
+    def __init__(self, option):
+        self.option = option
+
+    def __str__(self):
+        nbytes = self.option.nbytes()
+        test_data = generate_test_data(nbytes)
+        test_cases = "\n".join(
+            [
+                f"  TEST_CASE({i+2}, t0, 0x{test_data[i]:x}, ld t0, 0(a1); addi a1, a1, 8)"
+                for i in range(nbytes // 8)
+            ]
+        )
+        test_data_str = "\n".join([f"  .quad 0x{e:x}" for e in test_data])
+        return (TEMPLATE_HEADER + VLRE_TEMPLATE).format(
+            filename=self.option.filename,
+            inst_name=f"vl{self.option.nf}re{self.option.eew}.v",
+            extras="",
+            nf=self.option.nf,
+            nbytes=nbytes,
+            eew=self.option.eew,
+            test_cases=test_cases,
+            test_data_str=test_data_str,
+        )
+
+
+class Vlse(CodegenBase):
+    """Generate vle<EEW>.v, vse<EEW>.v tests."""
+
+    class Option(CodegenBase.Option):
+        """Vlse option."""
+
+        def __init__(self, filename, inst, lmul, eew, vl):
+            super().__init__(filename)
+            self.inst = inst
+            self.lmul = lmul
+            self.eew = eew
+            self.vl = vl
+
+        def nbytes(self):
+            return (self.vl * self.eew) // 8
+
+    def __init__(self, option):
+        self.option = option
+
+    def __str__(self):
+        nbytes = self.option.nbytes()
+        nquads = nbytes // 8
+        remaining_bytes = nbytes % 8
+        test_data_bytes = (VLEN * self.option.lmul) // 8 + 8
+        test_data = generate_test_data(test_data_bytes)
+        test_data_str = "\n".join([f"  .quad 0x{e:x}" for e in test_data])
+
+        start = inc(2)
```

Comment:
Tests cases are registered from case#3, Isn't it supposed to be like `start = inc(2)` to generate proper number of cases?